---
title: "A Comparison of GMAC and ADDIS"
author: 
-  Jarred Kvamme, University of Idaho
date: "9/22/2021"
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=8)
library(latex2exp)
library(gridExtra)
```

\section*{1. Overview}
Both MRPC-LOND and MRPC-ADDIS techniques inferred a large number of trans mediated trios. The trans-mediation model has been previously identified, but is not the commonly ackowledged mode of mediation. Since this result is surprising relative to the existing literature, we sought to apply another method for inferring mediation on a subset of GTEx trios analyzed herein by MRPC. The Genomic Mediation analysis with Adaptive Confounding (GMAC) algorithm allows for a unique selection of a subset of potential confounders, ${\bf X_{ij}}$ from a larger covariate pool, ${\bf H}$, for each trio. By taking advantage of the Principle of Mendelian Randomization, the authors filter ${\bf H}$ by removing common child and intermediate confounding variables (e.g variables associated with the eQTL as well as the cis/trans genes). Post-filtering, GMAC preforms a mediation test on the edge between the cis gene and trans gene via the regression of the trans-gene $T_j$ on the cis-eQTL $L_i$, cis-gene $C_i$, and the set of adaptively selected confounders ${\bf X_{ij}}$:

\begin{eqnarray} T_j = \beta_0 + \beta_1 C_i + \beta_2 L_i + {\bf \Gamma} {\bf X}_{ij} + \epsilon \end{eqnarray}

The mediation statistic is the observed $t$-value of the cis-gene coefficient $\beta_1$. A null distribution for no-mediation is constructed by iteratively permuting the values of the cis-transcript within each genotype and repeating the above regression. The authors argue that the permutation of the cis-transcript within the genotypes of the cis-eQTL removes the association between the cis and trans gene transcripts while preserving the higher order associations with the cis-eQTL. The resulting permutation test for mediation compares the observed relationship between the trans and cis gene to a null distribution constructed from a model with no association and assuming that possible confounding has been well adjusted via the selected covariates. 

It is important to note that the above mediation test describes only the association between cis-gene and trans-gene transcripts ($C_i \leftrightarrow T_j$ ) and does not consider possible effects between the cis-eQTL and the cis-gene transcript ($L_i \rightarrow C_i$), or the cis-eQTL and trans-gene transcript ($L_i \rightarrow T_j$).

\section*{2. Methods}
\subsection*{2.1 Applying GMAC to GTEx Trios}

To compare the GMAC and MRPC algorithms, we applied the GMAC algorithm to the top five GTEx tissues by sample size. Following with the creators of GMAC, we used the full set of principle components retained from the PCA of the expression matrix as the covariate pool, and three additional known confounders: the PCR used, the platform used, and sex of the individual in each sample [@yang2017identifying].    

Consistent with @yang2017identifying, the analysis was preformed using a common child and intermediate variable filtering  FDR of $10\%$ and a confounder selection FDR of $5\%$ for each trio. Each trio supplied to GMAC consisted of the cis-QTL and the PEER normalized cis and trans gene transcripts with the highest association to the eQTL. To mitigate missing values in the eQTL matrix, multiple imputation of the matrix of unique cis-eQTLs was preformed via multiple correspondence analysis (MCA) prior to its use in GMAC [@josse2016missmda]. The analysis was preformed twice on each trio, first with the cis gene as the mediator and second with the trans gene as the mediator. This allowed for GMAC inferred trios to be decomposed into the three groupings used under MRPC: 1) Cis-gene mediation, 2) Trans-gene mediation, 3) both (undirected). 

\subsection*{2.2 Comparison of GMAC and MRPC Results}
   
After applying GMAC to each tissue, the false discovery rate among the retained mediation p-values was controlled at the more liberal rate of $10\%$ [@yang2017identifying]. Each trio determined to have significant mediation after FDR filtering was compared with the regulatory network type inferred by MRPC-ADDIS. MRPC-ADDIS can infer three types of regulatory networks that contain an edge between the cis and trans gene (M1, M2, or M4). Since GMAC considers only the presence of the edge and not its direction, trios inferred to be one of M1, M2, or M4 under ADDIS, that were also significant under GMAC, were considered consistent (e.g $C_i \rightarrow T_j$; $T_j \rightarrow C_i$; $C_i \leftrightarrow T_j$ are synonymous under GMAC).   

\subsection*{2.3 Simulations}

\indent $\bullet$ The Small True Model Simulation (STM)

To further understand the conflict in edge determination between MRPC-ADDIS and GMAC, we simulated the mediation test - the test for the $\beta_1$ coefficient in the presence of all adaptively selected confounders, ${\bf X}_{ij}$, as described by the regression in \textbf{eq (1)} - under two different scenarios. (i) To observe the predictive power of the mediation test when the trans gene comes from a set of explanatory variables smaller than those in ${\bf X}_{ij}$, We simulated the trans gene of each trio using the linear relationship:

\begin{eqnarray} T^{\ast}_{j} = \hat{\beta_0}+\hat{\beta_1}C_i + \hat{\beta_2}L_i + \hat{{\bf\Gamma}}_W {\bf W}_{ij} +\epsilon  \end{eqnarray}

where $T^{\ast}_{j}$ is the simulated trans gene, the coefficients are replaced by their estimates from the regression in \textbf{eq (1)}, the errors are $\epsilon \sim N(0, \hat{\sigma})$, and ${\bf W}_{ij}$ is a subset of confounders in ${\bf X}_{ij}$ representing the "highly" significant confounders from \textbf{eq (1)} ($p<0.001$). Note that if the GMAC inferred mediation type was trans gene mediation only then the cis gene was simulated and the mediation test was preformed on the $\beta_1$ coefficient from regression of the cis gene: $C_i = \beta_0 + \beta_1T_j + \beta_2L_i + {\bf \Gamma X}_{ij}+\epsilon$. 

We refer to the trans-gene generating function in \textbf{(2)} as the Small True Model (STM) as the simulated trans gene comes from a model that is a subset of the explanatory variables in the analysis model described in \textbf{(1)}. The simulated mediation test then replaces $T_j$ in \textbf{(1)} with the simulated trans gene $T^{\ast}_j$. Therefore, the simulated mediation test under the STM can be decomposed as the test on the $\beta_1$ coefficient obtained from the regression:

\begin{eqnarray} T^{\ast}_{j} = \beta_0+\beta_1C_i + \beta_2L_i + {\bf\Gamma}_{1,W} {\bf W}_{ij} + {\bf \Gamma}_{2,M}{\bf M}_{ij} +\epsilon \end{eqnarray}

where ${\bf M}_{ij}$ represents the additional confounders in ${\bf X}_{ij}$ that are not included in ${\bf W}_{ij}$ see \textbf{Figure 1}.  

\indent $\bullet$ The Large True Model Simulation (LTM)


Conversely, a second simulation model was implemented to observe the power of the mediation test when the generating model for the trans gene is larger than the model used to infer the mediation relationship. In this scenario, the trans gene is simulated via:

\begin{eqnarray} T^{\ast}_{j} = \hat{\beta}_0+\hat{\beta}_1C_i + \hat{\beta}_2L_i + \hat{{\bf \Gamma}}_V{\bf V}_{ij} +\epsilon  \end{eqnarray}

which we refer to as the Large True Model (LTM). Note that the coefficient estimates in \textbf{(4)} come from the regression of the original data on a larger set than in \textbf{eq. (1)}:

\begin{eqnarray} T_{j} = \beta_0+\beta_1C_i + \beta_2L_i + {\bf\Gamma} {\bf X}_{ij} + {\bf \Gamma}_{G}{\bf G}_{ij} +\epsilon \end{eqnarray}

where ${\bf G}_{ij}$ represents the additional explanatory variables in ${\bf V}_{ij}$ not in ${\bf X}_{ij}$ (see \textbf{Figure 1}). The additional variables represented by ${\bf G}_{ij}$ were randomly selected from the confounder pool ${\bf H}$ with equal probability. The four sets of $p$-values (GMAC mediation $p$-value, GMAC permutation $p$-value, the STM mediation $p$-value, and the LTM mediation $p$-value) were visually inspected to determine the effect of model mis-specification on the power of the mediation and permutation test(s).


\indent $\bullet$ True GMAC Model Simulation (TGM)

We preformed a third and final simulation of the trans gene using the analysis model in GMAC given by \textbf{(1)} as the trans gene generating process. We then applied the MRPC model 

\begin{eqnarray}  T_j = \beta_0 + \beta_1 C_i +\beta_2 L_i + {\bf \Gamma}_z {\bf Z}_{ij} \end{eqnarray}

where ${\bf Z}_{ij}$ is a subset of both ${bf X}_{ij}$ and ${W}_{ij}$ representing only the confounders selected under the MRPC methodology. The goal was to analyze the power of MRPC when infer the larger GMAC model under the assumption that the GMAC model is the true.  


\indent $\bullet$ GMAC Self Simulation (GSS)

We created a fourth simulation scenario to test the importance of permutation on the analysis result. Each of the above simualtions uses variations understanding the inferential power of MRPC/GMAC when we know we have the incorrect model. In this scenario we take a different approach and use the model suggested by GMAC for each trio given by \textbf{eqn. (1)} to generate the trans gene such that:

\[ T_j^{\ast} = \hat{\beta}_0 + \hat{\beta}_1 C_i + \hat{\beta}_2 L_i + \hat{{\bf \Gamma}} {\bf X}_{ij} + \epsilon  \]

where $\epsilon \sim N(0, \hat{\sigma})$. We the apply ${\bf 1}$ to each simulated trio under this scenario to obtain the parametric and permutation inferences when we know the analysis model is correct.  




\section*{3. Results}
\subsection*{3.1 Comparing GMAC and MRPC}


In light of the surprising number of trans-gene mediation trios inferred by MRPC, we sought to compare our results with GMAC by applying the GMAC method to the top five GTEx tissues by sample size. It is important to note that the test for mediation used by GMAC describes only the association between cis-gene and trans-gene transcripts ($C_i \leftrightarrow T_j$ ) and does not consider the possible effects between the cis-eQTL and the cis-gene ($L_i \rightarrow C_i$), or the cis-eQTL and trans-gene ($L_i \rightarrow T_j$). Therefore, since GMAC considers only the presence of the mediation edge, trios inferred to be one of M1, M2, or M4 under ADDIS, that were also significant under GMAC, were considered consistent (e.g $C_i \rightarrow T_j$; $T_j \rightarrow C_i$; $C_i \leftrightarrow T_j$ are synonymous under GMAC).

At the $10\%$ false discovery rate, GMAC identified $2,160$ trios with an edge between the cis and trans genes out of $55,446$ total trios tested across the five selected tissues: Adipose subcutaneous, Tibial artery, Muscle skeletal, Sun exposed skin, and Whole blood. Of the trios with mediation edges, 653 were identified as the cis gene mediating the trans gene, 245 as trans gene mediating the cis gene and 1,345 as both ($29.1\%, 10.9\%$, and $60\%$ respectively). As can be seen from \textbf{Table 2}, the consistency in inferred mediation edges between the two methods varied between $39\%$ and $50\%$ of the trios across tissues. 

To uncover the computational differences between MRPC and GMAC, we focused on trios with conflicting results between the two methods (trios inferred M0 or M3 under MRPC). The primary differences we observed between the two algorithms for these trios were that the inclusion of a larger set of confounding variables by GMAC often had the effect of strengthening the association between the cis and trans genes. That is, let ${\bf Z}_{ij}$ denote the set of confounding variables used under MRPC and ${\bf X}_{ij}$ the larger set used by GMAC such that the columns $\{z_{ij} \} \subset \{x_{ij}\}$. Because the confounding variables under GMAC are selected such that they have a significant association with either the cis trans trans genes, the partial correlation between the cis gene and trans gene tends to strengthen as the column dimension of ${\bf Z}_{ij}$ approaches the column dimension of ${\bf X}_{ij}$. The result for GMAC is that under the mediation test, relatively weak associations ($0.1 \leq \rho \leq 0.2$) can be deemed significant. Conversely, for MRPC, as the size of the network increases, the method becomes increasingly conservative. Therefore, when ${\bf Z}_{ij} = {\bf X}_{ij}$ MRPC tends to infer the null model unless the association between two nodes in the network is substantial.

\subsection*{3.2 Simulation Results}

To observe the predictive differences between the MRPC-ADDIS and GMAC methods, we simulated the GMAC mediation test under two models for the trans gene: (i) when the true model generating the trans gene is smaller than the analysis model (STM), and (ii) when the model generating the trans gene is larger than the analysis model (LTM) see section ${\bf 2.3}$ for details. The STM simulated mediation test from (i) represents the case when the analysis model includes additional confounders not involved in the true trans gene generating process (i.e when we include more confounders than needed). Alternatively, the LTM simulated mediation test from (ii) represents a scenario similar to the test preformed by MRPC-ADDIS where the number of confounders included in the mediation test is a subset of the confounders in the model generating the trans gene with the greatest contribution to the response (i.e when we don't include enough confounders). 

In general, over-specifying or under-specifying the model does little to change the inference of the mediation test (\textbf{Figure 4}) when the number of confounders is already large (relative to the MRPC model). However, 

The blue dots in \textbf{Figure 4} distinguish specific trios which contain an eQTL with a rare allele present in the sample (typically with an allele frequency $\approx 1\%$). 




\newpage
\section{Tables and Figures}



```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
cn=c("Adipose Subcutaneous","Artery Tibial","Muscle Skeletal","Skin Sun Exposed","Whole Blood")

rn=c("Min.", "1st Qu.", "Median", "Mean", "3rd Qu.", "Max.")
dat=matrix(0, nrow=6, ncol = 5)
dat[,1]=c(0.000000, 0.000000, 0.000000, 0.006365, 0.003442, 0.156627)
dat[,2]=c(0.000000, 0.000000, 0.000000, 0.006625, 0.003425, 0.159247)
dat[,3]=c(0.000000, 0.000000, 0.000000, 0.006560, 0.002833, 0.158640)
dat[,4]=c(0.000000, 0.000000, 0.000000, 0.006701, 0.003306, 0.160331)
dat[,5]=c(0.000000, 0.000000, 0.000000, 0.006103, 0.002985, 0.155224)
colnames(dat)=cn
row.names(dat)=rn

kable(dat, caption = "Descriptive statistics for the distribution of missing values across the eQTL's for each tissue used in GMAC" )

```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
table.final=read.csv("C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/summary.all.tissues.final.csv")

table.final[,9]=(table.final$M1+table.final$M2+table.final$M4)/table.final[,8]
colnames(table.final)=c("Tissue", paste0("M", c(0:4)), "Other", "Total GMAC Inferred", "Percentage In Common")

kable(table.final, caption = "The breakdown of unqiue trios with inferred significant cis or trans mediation under GMAC across their respective ADDIS inferred regulatory networks. The column \"Percentage In Common\" is the proportion of significant trios that also contained a mediation edge in the regulatory network inferred under MRPC-ADDIS", digits = 4)

write.csv(table.final, "C:/Users/Bruin/Documents/GitHub/MRPC_support/Manuscript/GMAC_MRPC_comparison_final.csv")
```



```{r, echo=FALSE, message=FALSE, warning=FALSE}

table.totals=read.csv("C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/final_result_table10000.csv")
table.unique=read.csv("C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/GMAC.unique.trios.breakdown.csv")
table3=cbind.data.frame(table.totals[,2],c(1:5), table.totals[,3:5], table.unique[,-1], rowSums(table.unique[,-1]))
colnames(table3)=c("Tissue", "Sample Size", "Tested Trios", "Cis-Gene Mediation Trios", "Trans-Gene Mediation Trios", "Unique Cis", "Unique Trans", "Unique Both", "Unique Total")
table3$`Sample Size`=c(581, 584, 706, 605, 670)
table3$Tissue=c("AdiposeSubcutaneous", "ArteryTibial", "MuscleSkeletal", "SunExposedSkin", "WholeBlood")
kable(table3, caption = "Breakdown of trios with inferred mediation under GMAC across both cis and trans mediation types. The column \"Unique Both\" represents the intersection of the columns \"Total Cis Mediated\" and \"Total Trans Mediated\" ")


```



![A visual depiction of the relationship between the simulation models for the trans gene and the regression used for the mediation test](C:\Users\Bruin\Documents\GitHub\MRPC_support\Manuscript\GMACwriteup_sim_diagram.png)


```{r, fig.cap="A density plot for the absolute partial correlations to demonstrate the strengthening of the adjusted relationship under the GMAC mediation model.", echo=FALSE, message=FALSE, warning=FALSE}
library(ggpubr)

table.pcorM3=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/trios.pcors.summary.tableM3.txt", sep = "\t", header = T)

table.pcorM0=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/trios.pcors.summary.tableM0.txt", sep = "\t", header = T)

pcors=c(table.pcorM3$GMAC.pcor, table.pcorM3$ADDIS.pcor)
labels=c(rep('GMAC', length(table.pcorM3$GMAC.pcor)), rep("ADDIS", length(table.pcorM3$ADDIS.pcor)))

dat1=cbind.data.frame(pcors=pcors, labels=labels)

ggplot(data=dat1, aes(x=abs(pcors), fill=labels))+geom_density(alpha=0.5)+
  labs(title = "Density of Absolute Partial Correlations",
       x = expression(abs(rho)))



```



## variance across tissues plots


```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE}
#sim=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/simulations_wholeblood.txt", sep="\t", header = T)
library(ggpubr)
library(prodlim)
sim=read.csv(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/TRIOS_imbalanced_genotypes_WB.csv", header = T)
sim2=read.csv(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/TRIOS_imbalanced_genotypes_AS.csv", 
               header = T)
sim3=rbind.data.frame(sim,sim2)


sim3$alternative=(sim3$het+2*sim3$ha)/(2*rowSums(sim3[,36:38]))
sim3$reference=1-sim3$alternative
genotypes=sim3[,36:38]/rowSums(sim3[,36:38])

#floor variables

sim3$Perm.p.GMAC.floored=sim3$Perm.p.GMAC
sim3$Perm.p.GMAC.floored=replace(sim3$Perm.p.GMAC.floored, which(sim3$Perm.p.GMAC.floored==0), 1e-05)

sim3$Perm.p.MRPC.floored=sim3$Perm.p.MRPC
sim3$Perm.p.MRPC.floored=replace(sim3$Perm.p.MRPC.floored, which(sim3$Perm.p.MRPC.floored==0), 1e-05)

#LTM-STM permutation pvalues 
sim3$STM.perm.p.floored=sim3$STM.perm.p
sim3$STM.perm.p.floored=replace(sim3$STM.perm.p.floored, which(sim3$STM.perm.p.floored==0), 1e-05)

sim3$LTM.perm.p.floored=sim3$LTM.perm.p
sim3$LTM.perm.p.floored=replace(sim3$LTM.perm.p.floored, which(sim3$LTM.perm.p.floored==0), 1e-05)

sim3$TGM.perm.floored=sim3$TGM.perm.p
sim3$TGM.perm.floored=replace(sim3$TGM.perm.floored, which(sim3$TGM.perm.floored==0), 1e-05)

sim3$GSS.perm.floored=sim3$GSS.perm.p
sim3$GSS.perm.floored=replace(sim3$GSS.perm.floored, which(sim3$GSS.perm.floored==0), 1e-05)

#create color for allele frequencies of alternative or reference < 5% 
cv1=rep(0, dim(sim3)[1])
cv2=replace(cv1, which(sim3$alternative<=0.05 | sim3$reference<=0.05), 1)
h=0.05
sim3$colvar=as.factor(cv2)

A=ggplot(data = sim3, aes(x=Num.pcs.GMAC, y=Per.var.GMAC, color=Tissue))+
  geom_point()+
  ggtitle("%Var Accounted For by GMAC included PC's", subtitle = "GMAC")+
  xlab("Number of PC's")+
  ylab("%Total Variance In Expression")


B=ggplot(data = sim3, aes(x=Num.pcs.MRPC, y=Per.var.MRPC, color=Tissue))+
  geom_point()+
  ggtitle("%Var Accounted For by MRPC included PC's", subtitle = "MRPC")+
  xlab("Number of PC's")+
  ylab("%Total Variance In Expression")


ggarrange(A, B, nrow = 2, labels = c("A", "B"))




```




## LTM and STM plots


```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE, results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim3)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(STM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('sim. Parametric P-value')),
       x = expression(log10('GMAC Obs. Parametric P-value')),
       subtitle = "Small True Model Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

B=ggplot(data = sim3)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(LTM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('GMAC Obs Parametric P-value')),
       subtitle = "Large True Model Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

A1=ggplot(data = sim3)+
   geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(STM.perm.p.floored), color=colvar))+
   theme(legend.position = "none")+
   labs(y = expression(log10('Sim. Permutation P-value')),
        x = expression(log10('GMAC Obs. Perm. P-value')),
        subtitle = "Small True Model Simulation")+
   scale_x_continuous(limits = c(-5,0))+
   scale_y_continuous(limits = c(-5,0))+
   geom_abline(intercept = 0, slope = 1, linetype="dotted")+
   geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
   geom_hline(yintercept = log10(h), color="red", linetype="dashed")


B1=ggplot(data = sim3)+
   geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(LTM.perm.p.floored), color=colvar))+
   theme(legend.position = "none")+
   labs(y = expression(log10('Sim. Permutation P-value')),
        x = expression(log10('GMAC Obs. Perm. P-value')),
        subtitle = "Large True Model Simulation")+
   scale_x_continuous(limits = c(-5,0))+
   scale_y_continuous(limits = c(-5,0))+
   geom_abline(intercept = 0, slope = 1, linetype="dotted")+
   geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
   geom_hline(yintercept = log10(h), color="red", linetype="dashed")


ggarrange(A, B, A1, B1, labels = c("A","B","C","D") , nrow=2, ncol=2)

```














## TGM and MRPC plots

```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim3)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(TGM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('Obs. GMAC Parametric P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")



B=ggplot(data = sim3)+
  geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(TGM.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Permutation P-value')),
       x = expression(log10('GMAC Obs. Perm. P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-5,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim3)+
  geom_point(aes(x=log10(TGM.med.p), y=log10(TGM.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. TGM Perm. P-value')),
       x = expression(log10('Sim. TGM Para. P-value')),
       subtitle = "True GMAC Model Simulation Perm. Vs. Para")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-6,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

D=ggplot(data = sim3)+
  geom_point(aes(x=log10(Med.pvalue.MRPC), y=log10(Perm.p.MRPC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('MRPC Permutation P-value')),
       x = expression(log10('MRPC Parametric P-value')),
       subtitle = "MRPC Permutation Vs. Parametric")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(A, B, C, D, labels = c("A","B","C","D") , nrow=2, ncol=2)

```






## GSS and GMAC plots 

```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim3)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(GSS.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('Obs. GMAC Parametric P-value')),
       subtitle = "GMAC Self Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")



B=ggplot(data = sim3)+
  geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(GSS.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Permutation P-value')),
       x = expression(log10('GMAC Obs. Perm. P-value')),
       subtitle = "GMAC self Simulation")+
  scale_x_continuous(limits = c(-5,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim3)+
  geom_point(aes(x=log10(GSS.med.p), y=log10(GSS.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Perm. P-value')),
       x = expression(log10('Sim. Para. P-value')),
       subtitle = "GMAC Self Simulation Perm. Vs. Para")+
  scale_x_continuous(limits = c(-5,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

D=ggplot(data = sim3)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(Perm.p.GMAC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('GMAC Permutation P-value')),
       x = expression(log10('GMAC Parametric P-value')),
       subtitle = "GMAC Permutation Vs. Parametric")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(A, B, C, D, labels = c("A","B","C","D") , nrow=2, ncol=2)

```


## Nominal P-value plots



```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.STM), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "Small True Model Simulation")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-16,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")



B=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.LTM), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "Large True Model Simulation")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-16,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.TGM), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-16,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

D=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.GSS), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "GMAC Self Simulation")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-16,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

E=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.MRPC), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Obs. MRPC Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "MRPC vs. GMAC")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-16,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

G=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Perm.p.GMAC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Obs. Perm. P-value')),
       x = expression(log10('Obs. Nominal P-value')),
       subtitle = "GMAC Nominal vs Permutation")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(A, B, C, D, labels = c("A","B","C","D") , nrow=2, ncol=2)

```




```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

E=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.MRPC), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Obs. MRPC Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "MRPC vs. GMAC")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-16,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")
h=0.05
G=ggplot(data = sim3)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(sim3$Med.pvalue.GMAC), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Obs. Para. P-value')),
       x = expression(log10('Obs. Nominal P-value')),
       subtitle = "GMAC Nominal vs Pare")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(E, G, labels = c("E","G") , nrow=2, ncol=2)

```




## boxplots for PC inclusion differences


```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE, results='hide',fig.keep='all'}

sim3$STM.med.p.floored=sim3$STM.med.p
sim3$STM.med.p.floored=replace(sim3$STM.med.p.floored, which(sim3$STM.med.p.floored<=2e-16), 2e-16)

sim3$LTM.med.p.floored=sim3$LTM.med.p
sim3$LTM.med.p.floored=replace(sim3$LTM.med.p.floored, which(sim3$LTM.med.p.floored<=2e-16), 2e-16)

sim3$TGM.med.floored=sim3$TGM.med.p
sim3$TGM.med.floored=replace(sim3$TGM.med.floored, which(sim3$TGM.med.floored<=2e-16), 2e-16)

TT=summary(as.factor(sim3$STM.dim.diff))
rmv=as.numeric(names(TT[TT==1]))

sim.sub=sim3[-c(match(rmv, sim3$STM.dim.diff)),]

h=0.05
A=ggplot(data = sim.sub)+
  geom_boxplot(aes(x=as.factor(STM.dim.diff), y=log10(STM.med.p.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('STM Sim. Parametric P-value')),
       x = "Number of Extra Components In Analysis Model",
       subtitle = "Small True Model")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


B=ggplot(data = sim3)+
  geom_boxplot(aes(x=as.factor(LTM.dim.diff), y=log10(LTM.med.p.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('LTM Sim. Parametric P-value')),
       x = "Number of Missing Components In Analysis Model",
       subtitle = "Large True Model")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim3)+
  geom_boxplot(aes(x=as.factor(Num.pcs.GMAC-Num.pcs.MRPC), y=log10(TGM.med.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('TGM Sim. Parametric P-value')),
       x = "Number of Missing Components In Analysis Model",
       subtitle = "True GMAC Model Simulation")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")




ggarrange(A, B, C, labels = c("A","B","C") , nrow=3)

```











## variation differences between sig and non-sig trios GMAC/MRPC



```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}




set.seed(222)


h=0.05

sim.sub1=subset(sim3, Med.pvalue.GMAC<0.05 & Med.pvalue.MRPC>0.05)
sim.sub1$`Model Significance`=rep("GMAC p < 0.05 vs MRPC p > 0.05", dim(sim.sub1)[1])
sim.sub2=subset(sim3, Med.pvalue.GMAC>0.05 & Med.pvalue.MRPC>0.05)
sim.sub2$`Model Significance`=rep("GMAC p > 0.05 vs MRPC p > 0.05", dim(sim.sub2)[1])

sim.sub=rbind.data.frame(sim.sub1, sim.sub2)
sim.sub$`Model Significance`=as.factor(sim.sub$`Model Significance`)

A=ggplot(data = sim.sub)+
  geom_boxplot(aes(x=`Model Significance`, y=Per.var.GMAC-Per.var.MRPC, fill=`Model Significance`))+
  theme(axis.text.x=element_blank())+
  labs(y="Diff. In Cumulative Variation In Included PC's")

mean1=mean(sim.sub1$Per.var.GMAC-sim.sub1$Per.var.MRPC)
mean2=mean(sim.sub2$Per.var.GMAC-sim.sub2$Per.var.MRPC)

B=ggplot(data = sim.sub)+geom_density(aes(x=Per.var.GMAC-Per.var.MRPC, fill=`Model Significance`), alpha=0.3, show.legend = FALSE)+
  geom_vline(xintercept = mean1, linetype="dotted")+
  geom_vline(xintercept = mean2, linetype="dotted")+
  theme_minimal()+
  labs(x="Diff. In Cumulative Variation In Included PC's")


C=ggplot(data = sim3)+
  geom_point(aes(x=GMAC.Nominal.p, y=Nominal.p.MRPC, color=colvar, label=Trio.Num))+
  theme(legend.position = "none")+
  labs(y = "Observed Nominal P-value (MRPC)",
       x = 'Observed Nominal P-value (GMAC)',
       subtitle = "MRPC vs. GMAC:  Color = Rare Variants")+
  geom_hline(yintercept = h, linetype="dashed", color="red")+
  #geom_vline(xintercept = h, linetype="dashed", color="red")+
  scale_y_continuous(limits = c(0, h))+
  geom_abline(slope = 1, intercept = 0, linetype="dashed")

MRPC.class=as.factor(sim3$ADDIS.inf.Class)
D=ggplot(data = sim3)+
  geom_point(aes(x=GMAC.Nominal.p, y=Nominal.p.MRPC, color=MRPC.class))+
  #theme(legend.position = "none")+
  labs(y = "Obs. Nominal P-value (MRPC)",
       x = 'Obs. Nominal P-value (GMAC)',
       subtitle = "MRPC vs. GMAC:  Color = Inf. Model Type")+
  geom_hline(yintercept = h, linetype="dashed", color="red")+
  geom_text(aes( -0.0004, h, label = h, vjust = -1), size = 5)+
  scale_y_continuous(limits = c(0, h))+
  geom_abline(slope = 1, intercept = 0, linetype="dashed")



ggarrange(A,B,C,D, labels = c("A","B","C","D"), nrow=2, ncol=2)


```











$\bullet$ Rare allele plots


```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE}

trio9AS=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/trio9AS.txt", sep="\t", header = T)


ggplot(data = trio9AS)+
  geom_point(aes(x=cis.gene, y=trans.gene, color=as.factor(SNP)), size=3)+
  labs(x = expression("Cis Gene: ENSG00000163682"),
       y = expression("Trans Gene: ENSG00000237550"),
       col = expression('Genotype Of \nSubject'),
       subtitle = "SNP ID: chr4_39448204_G_A_b38")+theme_pubr()

```



```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE}

trio1922ArT=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/trio1922ArT.txt", sep="\t", header = T)
pdf(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/Manuscript/SNP_effect_example_trio.pdf")
A=ggplot(data = trio1922ArT)+
  geom_point(aes(x=cis.gene, y=trans.gene, color=as.factor(SNP)), size=3)+
  labs(x = expression("Cis Gene: ENSG00000167526"),
       y = expression("Trans Gene: ENSG00000215030"),
       col = expression('Genotype Of \nSubject'),
       subtitle = "SNP ID: chr16_89561052_T_G_b38")+theme_pubr()
plot(A)
dev.off()
plot(A)
```




```{r, fig.cap="\\textbf{A} and \\textbf{C}: Example scatter plots of trios from subcutaneous adipose tissue and whole blood (respectively) with a rare allele present in the sample: Note that 0 indicates individuals homozygous for the reference allele, 1 indicates hetezygous individuals and 2 indicates individuals who are homozygous for the alternative (rare) allele. The apparent outlier represents a single individual in the sample who was homozygous for the rare allele, and \\textbf{B} and \\textbf{C} are the scatter plots with the point(s) for the homozygous-alternative individuals removed and a confidence ellipse calculated over the remaining homozygous-reference and heterzygous individals.", echo=FALSE, message=FALSE, warning=FALSE}

trio5683AS=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/Trio5683AS.txt", sep="\t", header = T)
trio8112WB=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/Trio8112WB.txt", sep="\t", header = T)

pdf("C:/Users/Bruin/Documents/GitHub/MRPC_support/Manuscript/SNP_effect_example_trio2.pdf")
A=ggplot(data = trio5683AS)+
  geom_point(aes(x=cis.gene, y=trans.gene, color=as.factor(SNP)))+
  labs(x = expression("Cis Gene: ENSG00000145715"),
       y = expression("Trans Gene: ENSG00000264330"),
       col = expression('Genotype Of \nSubject'),
       subtitle = "SNP ID: chr5_86411404_G_A_b38")



B=ggplot(data = subset(trio5683AS, SNP!=2))+
  geom_point(aes(x=cis.gene, y=trans.gene, color=as.factor(SNP)))+
  stat_ellipse(geom="polygon",aes(x=cis.gene, y=trans.gene, fill=as.factor(SNP)),alpha=0.35, show.legend=F)+
  labs(x = expression("Cis Gene: ENSG00000145715"),
       y = expression("Trans Gene: ENSG00000264330"),
       col = expression('Genotype Of \nSubject'),
       subtitle = "SNP ID: chr5_86411404_G_A_b38")
ggarrange(A,B, labels = c("A","B"), nrow=2, ncol=1)
dev.off()


C=ggplot(data = trio8112WB)+
  geom_point(aes(x=cis.gene, y=trans.gene, color=as.factor(SNP)))+
  labs(x = expression("Cis Gene: ENSG00000272356"),
       y = expression("Trans Gene: ENSG00000238423"),
       col = expression('Genotype Of \nSubject'),
       subtitle = "SNP ID: chr6_111433344_G_A_b38")


D=ggplot(data = subset(trio8112WB, SNP!=2))+
  geom_point(aes(x=cis.gene, y=trans.gene, color=as.factor(SNP)))+
  stat_ellipse(geom="polygon",aes(x=cis.gene, y=trans.gene, fill=as.factor(SNP)),alpha=0.35, show.legend=F)+
  labs(x = expression("Cis Gene: ENSG00000272356"),
       y = expression("Trans Gene: ENSG00000238423"),
       col = expression('Genotype Of \nSubject'),
       subtitle = "SNP ID: chr6_111433344_G_A_b38")


ggarrange(A,B,C,D, labels = c("A","B","C","D"), nrow=2, ncol=2)

```


























\newpage
\section*{References}