---
title: "GMACwriteup2.RMD"
author: 
-  Jarred Kvamme, University of Idaho
date: "9/22/2021"
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=8)
library(latex2exp)
library(gridExtra)
```




```{r, fig.cap="(A) scatter plots for the number of PC's included by GMAC and MRPC (B) and the percentage of total variation in expression summarized by the included PC's for each model", echo=FALSE, message=FALSE, warning=FALSE}
#sim=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/simulations_wholeblood.txt", sep="\t", header = T)
library(ggpubr)
library(prodlim)
sim=read.csv(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/TRIOS_imbalanced_genotypes_WB_m0m3.csv", header = T)
sim2=read.csv(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/TRIOS_imbalanced_genotypes_AS_m0m3.csv", header = T)

sim3=rbind.data.frame(sim,sim2)


sim$alternative=(sim$het+2*sim$ha)/(2*rowSums(sim[,26:28]))
sim$reference=1-sim$alternative
genotypes=sim[,26:28]/rowSums(sim[,26:28])

#floor variables

sim$Perm.p.GMAC.floored=sim$Perm.p.GMAC
sim$Perm.p.GMAC.floored=replace(sim$Perm.p.GMAC.floored, which(sim$Perm.p.GMAC.floored<=2e-16), 2e-16)

sim$Perm.p.MRPC.floored=sim$Perm.p.MRPC
sim$Perm.p.MRPC.floored=replace(sim$Perm.p.MRPC.floored, which(sim$Perm.p.MRPC.floored<=2e-16), 2e-16)

#LTM-STM permutation pvalues 
sim$STM.perm.p.floored=sim$STM.perm.p
sim$STM.perm.p.floored=replace(sim$STM.perm.p.floored, which(sim$STM.perm.p.floored<=2e-16), 2e-16)

sim$LTM.perm.p.floored=sim$LTM.perm.p
sim$LTM.perm.p.floored=replace(sim$LTM.perm.p.floored, which(sim$LTM.perm.p.floored<=2e-16), 2e-16)

sim$TGM.perm.floored=sim$TGM.perm
sim$TGM.perm.floored=replace(sim$TGM.perm.floored, which(sim$TGM.perm.floored<=2e-16), 2e-16)

#create color for allele frequencies of alternative or reference < 5% 
cv1=rep(0, dim(sim)[1])
cv2=replace(cv1, which(sim$alternative<=0.05 | sim$reference<=0.05), 1)
h=0.05
sim$colvar=as.factor(cv2)

A=ggplot(data = sim3, aes(x=Num.pcs.GMAC, y=Per.var.GMAC, color=Tissue))+
  geom_point()+
  ggtitle("%Var Accounted For by GMAC included PC's", subtitle = "GMAC")+
  xlab("Number of PC's")+
  ylab("%Total Variance In Expression")

B=ggplot(data = sim3, aes(x=Num.pcs.MRPC, y=Per.var.MRPC, color=Tissue))+
  geom_point()+
  ggtitle("%Var Accounted For by MRPC included PC's", subtitle = "MRPC")+
  xlab("Number of PC's")+
  ylab("%Total Variance In Expression")

ggarrange(A, B, nrow = 2, labels = c("A", "B"))




```




```{r, fig.cap="\\textbf{(A)} and \\textbf{(B)} the relationship between the observed $p$-value of the mediation test and the $p$-value from the simulated mediation test under the STM and LTM scenarios. \\textbf{(C)} and \\textbf{(D)} the relationship between the observed $p$-value of the permuted mediation test and the $p$-value from the simulated mediation tests under the STM and LTM scenarios. In all plots blue points represent trios with a rare allele observed at the SNP loci e.g having a frequency among subjects below $5\\%$. The dotted lines define the zone of insignificance at $\\log_{10}(\\alpha) = -1.301$ scale. Note that $p$-values < 2e-16 have been floored to $2e-16$", echo=FALSE, message=FALSE, warning=FALSE, results='hide',fig.keep='all'}


#set.seed(222)


h=0.05
A=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(STM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('GMAC Obs. Parametric P-value')),
       subtitle = "Small True Model Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


B=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(LTM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('GMAC Obs Parametric P-value')),
       subtitle = "Large True Model Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

A1=ggplot(data = sim)+
  geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(STM.perm.p.floored), color=colvar))+
  theme(legend.position = "none")+
   labs(y = expression(log10('Sim. Permutation P-value')),
        x = expression(log10('GMAC Obs. Permutation P-value')),
        subtitle = "Small True Model Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

B1=ggplot(data = sim)+
  geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(LTM.perm.p.floored), color=colvar))+
  theme(legend.position = "none")+
   labs(y = expression(log10('Sim. Permutation P-value')),
        x = expression(log10('GMAC Obs. Permutation P-value')),
        subtitle = "Large True Model Simulation")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


ggarrange(A, B, A1, B1, labels = c("A","B","C","D") , nrow=2, ncol=2)

```
















```{r, fig.cap="\\textbf{A and B} Comparison of the observed parametric and permuation $p$-values to those simulated under the True GMAC model scenario. \\textbf{C} comparison of the parametric to permuation $p$-values withing the TGM simulation. \\textbf{D} comparison of the parametric and permuation $p$-values within MRPC", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}


#set.seed(222)



h=0.05
A=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(TGM.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. TGM Parametric P-value')),
       x = expression(log10('Obs. GMAC Parametric P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")




B=ggplot(data = sim)+
  geom_point(aes(y=log10(TGM.perm.floored), x=log10(Perm.p.GMAC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. TGM Permutation P-value')),
       x = expression(log10('Obs. GMAC Permutation P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim)+
  geom_point(aes(x=log10(TGM.p), y=log10(TGM.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. TGM Perm. P-value')),
       x = expression(log10('Sim. TGM Para. P-value')),
       subtitle = "True GMAC Model Simulation Perm. Vs. Para")+
  scale_x_continuous(limits = c(-15,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

D=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.MRPC), y=log10(Perm.p.MRPC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('MRPC Permutation P-value')),
       x = expression(log10('MRPC Parametric P-value')),
       subtitle = "MRPC Permutation Vs. Parametric")+
  scale_x_continuous(limits = c(-7,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(A, B, C, D, labels = c("A","B","C","D") , nrow=2, ncol=2)

```








```{r, fig.cap="\\textbf{(A)} and \\textbf{(B)} The relationship between the simulated mediation $p$-value and the number of additional (STM scenario) or missing (LTM scenario) confounders in the analysis model relative to the true model generating the trans gene. In both plots blue boxes represent trios with a rare allele observed at the SNP loci e.g having a frequency among subjects below $5\\%$.", echo=FALSE, message=FALSE, warning=FALSE, results='hide',fig.keep='all'}



sim$STM.med.p.floored=sim$STM.med.p
sim$STM.med.p.floored=replace(sim$STM.med.p.floored, which(sim$STM.med.p.floored<=2e-16), 2e-16)

sim$LTM.med.p.floored=sim$LTM.med.p
sim$LTM.med.p.floored=replace(sim$LTM.med.p.floored, which(sim$LTM.med.p.floored<=2e-16), 2e-16)

sim$TGM.med.floored=sim$TGM.p
sim$TGM.med.floored=replace(sim$TGM.med.floored, which(sim$TGM.med.floored<=2e-16), 2e-16)

TT=summary(as.factor(sim$STM.dim.diff))
rmv=as.numeric(names(TT[TT==1]))

sim.sub=sim[-c(match(rmv, sim$STM.dim.diff)),]

h=0.05
A=ggplot(data = sim.sub)+
  geom_boxplot(aes(x=as.factor(STM.dim.diff), y=log10(STM.med.p.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('STM Sim. Parametric P-value')),
       x = "Number of Extra Components In Analysis Model",
       subtitle = "Small True Model")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


B=ggplot(data = sim)+
  geom_boxplot(aes(x=as.factor(LTM.dim.diff), y=log10(LTM.med.p.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('LTM Sim. Parametric P-value')),
       x = "Number of Missing Components In Analysis Model",
       subtitle = "Large True Model")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim)+
  geom_boxplot(aes(x=as.factor(Num.pcs.GMAC-Num.pcs.MRPC), y=log10(TGM.med.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('TGM Sim. Parametric P-value')),
       x = "Number of Missing Components In Analysis Model",
       subtitle = "True GMAC Model")+

  geom_hline(yintercept = log10(h), color="red", linetype="dashed")




ggarrange(A, B, C, labels = c("A","B","C") , nrow=3)


```





























```{r, fig.cap="\\textbf{A} The distribution of the difference between the GMAC and MRPC models in terms of in percentage of variation captured by the PCs/confounders included by each model, and under the special case of trios for which the mediation test was significant under GMAC but insignificant under MRPC \\textbf{B} The distribution described for \\textbf{A} but under the alternative scenario when both GMAC and MRPC mediation tests were insignificant (not to be confused with the permutation test which is significant for GMAC). \\textbf{C} Comparison of the permutation and mediation $p$-values under GMAC. Blue points indicate trios with a corresponding rare allele at the SNP loci. \\textbf{D} Comparison of the permutation and mediation $p$-values under GMAC and excluding trios with a corresponding rare allele for visibility. In \\textbf{A,B,D} points along the black diagonal lines indicate near $1:1$ correspondence in the $p$-values between methods.", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}




set.seed(222)


h=0.05

sim.sub1=subset(sim, Med.pvalue.GMAC<0.05 & Med.pvalue.MRPC>0.05)
sim.sub1$`Model Significance`=rep("GMAC p < 0.05 vs MRPC p > 0.05", dim(sim.sub1)[1])
sim.sub2=subset(sim, Med.pvalue.GMAC>0.05 & Med.pvalue.MRPC>0.05)
sim.sub2$`Model Significance`=rep("GMAC p > 0.05 vs MRPC p > 0.05", dim(sim.sub2)[1])

sim.sub=rbind.data.frame(sim.sub1, sim.sub2)
sim.sub$`Model Significance`=as.factor(sim.sub$`Model Significance`)

A=ggplot(data = sim.sub)+
  geom_boxplot(aes(x=`Model Significance`, y=Per.var.GMAC-Per.var.MRPC, fill=`Model Significance`))+
  theme(axis.text.x=element_blank())+
  labs(y="Diff. In Cumulative Variation In Included PC's")

mean1=mean(sim.sub1$Per.var.GMAC-sim.sub1$Per.var.MRPC)
mean2=mean(sim.sub2$Per.var.GMAC-sim.sub2$Per.var.MRPC)

B=ggplot(data = sim.sub)+geom_density(aes(x=Per.var.GMAC-Per.var.MRPC, fill=`Model Significance`), alpha=0.3, show.legend = FALSE)+
  geom_vline(xintercept = mean1, linetype="dotted")+
  geom_vline(xintercept = mean2, linetype="dotted")+
  theme_minimal()+
  labs(x = "Diff. In Cumulative Variation In Included PC's")


C=ggplot(data = sim)+
  geom_point(aes(x=Med.pvalue.GMAC, y=Med.pvalue.MRPC, color=colvar))+
  #geom_point(aes(x=Med.pvalue.GMAC, y=STM.med.p))+
  theme(legend.position = "none")+
  #ggtitle("Scatter of Mediation Pvalues", )+
    labs(y = "Observed Parametric P-value (MRPC)",
       x = 'Observed Parametric P-value (GMAC)',
       subtitle = "Comparing Paramteric Tests Under GMAC and MRPC")+
  geom_hline(yintercept = h, linetype="dashed", color="red")+
  geom_vline(xintercept = h, linetype="dashed", color="red")+
  geom_abline(slope = 1, intercept = 0, linetype="dashed")

  
D=ggplot(data = sim)+
  geom_point(aes(x=Perm.p.GMAC.floored, y=Perm.p.MRPC.floored, color=colvar))+
  #geom_point(aes(x=Med.pvalue.GMAC, y=STM.med.p))+
  theme(legend.position = "none")+
  #ggtitle("Scatter of Mediation Pvalues", )+
    labs(y = "Observed Permutation P-value (MRPC)",
       x = 'Observed Permutation P-value (GMAC)',
       subtitle = "Comparing Permutation Tests Under GMAC and MRPC")+
  geom_hline(yintercept = h, linetype="dashed", color="red")+
  geom_text(aes( -0.0004, h, label = h, vjust = -1), size = 5)+
  scale_y_continuous(limits = c(0, h))+
  geom_abline(slope = 1, intercept = 0, linetype="dashed")



ggarrange(A,B,C,D, labels = c("A","B","C","D"), nrow=2, ncol=2)


```