---
title: "GMACwriteup2.RMD"
author: 
-  Jarred Kvamme, University of Idaho
date: "9/22/2021"
output:
  pdf_document:
    fig_caption: yes
    keep_tex: yes
bibliography: references.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=8)
library(latex2exp)
library(gridExtra)
```



## variance across tissues plots


```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE}
#sim=read.table(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/simulations_wholeblood.txt", sep="\t", header = T)
library(ggpubr)
library(prodlim)
sim=read.csv(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/TRIOS_imbalanced_genotypes_WB_m0m3.csv", header = T)
sim2=read.csv(file="C:/Users/Bruin/Documents/GitHub/MRPC_support/GMAC analysis/master_tables/TRIOS_imbalanced_genotypes_AS_m0m3.csv", header = T)

sim3=rbind.data.frame(sim,sim2)


sim$alternative=(sim$het+2*sim$ha)/(2*rowSums(sim[,36:38]))
sim$reference=1-sim$alternative
genotypes=sim[,36:38]/rowSums(sim[,36:38])

#floor variables

sim$Perm.p.GMAC.floored=sim$Perm.p.GMAC
sim$Perm.p.GMAC.floored=replace(sim$Perm.p.GMAC.floored, which(sim$Perm.p.GMAC.floored==0), 1e-05)

sim$Perm.p.MRPC.floored=sim$Perm.p.MRPC
sim$Perm.p.MRPC.floored=replace(sim$Perm.p.MRPC.floored, which(sim$Perm.p.MRPC.floored==0), 1e-05)

#LTM-STM permutation pvalues 
sim$STM.perm.p.floored=sim$STM.perm.p
sim$STM.perm.p.floored=replace(sim$STM.perm.p.floored, which(sim$STM.perm.p.floored==0), 1e-05)

sim$LTM.perm.p.floored=sim$LTM.perm.p
sim$LTM.perm.p.floored=replace(sim$LTM.perm.p.floored, which(sim$LTM.perm.p.floored==0), 1e-05)

sim$TGM.perm.floored=sim$TGM.perm
sim$TGM.perm.floored=replace(sim$TGM.perm.floored, which(sim$TGM.perm.floored==0), 1e-05)

sim$GSS.perm.floored=sim$GSS.perm.p
sim$GSS.perm.floored=replace(sim$GSS.perm.floored, which(sim$GSS.perm.floored==0), 1e-05)

#create color for allele frequencies of alternative or reference < 5% 
cv1=rep(0, dim(sim)[1])
cv2=replace(cv1, which(sim$alternative<=0.05 | sim$reference<=0.05), 1)
h=0.05
sim$colvar=as.factor(cv2)

A=ggplot(data = sim3, aes(x=Num.pcs.GMAC, y=Per.var.GMAC, color=Tissue))+
  geom_point()+
  ggtitle("%Var Accounted For by GMAC included PC's", subtitle = "GMAC")+
  xlab("Number of PC's")+
  ylab("%Total Variance In Expression")

B=ggplot(data = sim3, aes(x=Num.pcs.MRPC, y=Per.var.MRPC, color=Tissue))+
  geom_point()+
  ggtitle("%Var Accounted For by MRPC included PC's", subtitle = "MRPC")+
  xlab("Number of PC's")+
  ylab("%Total Variance In Expression")

ggarrange(A, B, nrow = 2, labels = c("A", "B"))




```


## LTM and STM plots


```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE, results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(STM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('GMAC Obs. Parametric P-value')),
       subtitle = "Small True Model Simulation")+
  scale_x_continuous(limits = c(-8,0))+
  scale_y_continuous(limits = c(-8,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

B=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(LTM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('GMAC Obs Parametric P-value')),
       subtitle = "Large True Model Simulation")+
  scale_x_continuous(limits = c(-8,0))+
  scale_y_continuous(limits = c(-8,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

A1=ggplot(data = sim)+
   geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(STM.perm.p.floored), color=colvar))+
   theme(legend.position = "none")+
   labs(y = expression(log10('Sim. Permutation P-value')),
        x = expression(log10('GMAC Obs. Perm. P-value')),
        subtitle = "Small True Model Simulation")+
   scale_x_continuous(limits = c(-5,0))+
   scale_y_continuous(limits = c(-5,0))+
   geom_abline(intercept = 0, slope = 1, linetype="dotted")+
   geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
   geom_hline(yintercept = log10(h), color="red", linetype="dashed")


B1=ggplot(data = sim)+
   geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(LTM.perm.p.floored), color=colvar))+
   theme(legend.position = "none")+
   labs(y = expression(log10('Sim. Permutation P-value')),
        x = expression(log10('GMAC Obs. Perm. P-value')),
        subtitle = "Large True Model Simulation")+
   scale_x_continuous(limits = c(-5,0))+
   scale_y_continuous(limits = c(-5,0))+
   geom_abline(intercept = 0, slope = 1, linetype="dotted")+
   geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
   geom_hline(yintercept = log10(h), color="red", linetype="dashed")


ggarrange(A, B, A1, B1, labels = c("A","B","C","D") , nrow=2, ncol=2)

```
















## TGM and MRPC plots

```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(TGM.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('Obs. GMAC Parametric P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-8,0))+
  scale_y_continuous(limits = c(-8,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")



B=ggplot(data = sim)+
  geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(TGM.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Permutation P-value')),
       x = expression(log10('GMAC Obs. Perm. P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-5,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim)+
  geom_point(aes(x=log10(TGM.med.p), y=log10(TGM.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. TGM Perm. P-value')),
       x = expression(log10('Sim. TGM Para. P-value')),
       subtitle = "True GMAC Model Simulation Perm. Vs. Para")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

D=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.MRPC), y=log10(Perm.p.MRPC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('MRPC Permutation P-value')),
       x = expression(log10('MRPC Parametric P-value')),
       subtitle = "MRPC Permutation Vs. Parametric")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(A, B, C, D, labels = c("A","B","C","D") , nrow=2, ncol=2)

```




## GSS and GMAC plots 

```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(GSS.med.p), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Parametric P-value')),
       x = expression(log10('Obs. GMAC Parametric P-value')),
       subtitle = "GMAC Self Simulation")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-10,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")



B=ggplot(data = sim)+
  geom_point(aes(x=log10(Perm.p.GMAC.floored), y=log10(GSS.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Permutation P-value')),
       x = expression(log10('GMAC Obs. Perm. P-value')),
       subtitle = "GMAC self Simulation")+
  scale_x_continuous(limits = c(-5,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim)+
  geom_point(aes(x=log10(GSS.med.p), y=log10(GSS.perm.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Perm. P-value')),
       x = expression(log10('Sim. Para. P-value')),
       subtitle = "GMAC Self Simulation Perm. Vs. Para")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

D=ggplot(data = sim)+
  geom_point(aes(x=log10(Med.pvalue.GMAC), y=log10(Perm.p.GMAC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('GMAC Permutation P-value')),
       x = expression(log10('GMAC Parametric P-value')),
       subtitle = "GMAC Permutation Vs. Parametric")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(A, B, C, D, labels = c("A","B","C","D") , nrow=2, ncol=2)

```










## Nominal P-value plots



```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

#set.seed(222)


h=0.05
A=ggplot(data = sim)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.STM), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "Small True Model Simulation")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-10,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")



B=ggplot(data = sim)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.LTM), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "Large True Model Simulation")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.TGM), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "True GMAC Model Simulation")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-10,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

D=ggplot(data = sim)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.GSS), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Sim. Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "GMAC Self Simulation")+
  scale_x_continuous(limits = c(-10,0))+
  scale_y_continuous(limits = c(-15,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")



ggarrange(A, B, C, D, labels = c("A","B","C","D") , nrow=2, ncol=2)

```






```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}

E=ggplot(data = sim)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Nominal.p.MRPC), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Obs. MRPC Nominal P-value')),
       x = expression(log10('Obs. GMAC Nominal P-value')),
       subtitle = "MRPC vs. GMAC")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-16,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

G=ggplot(data = sim)+
  geom_point(aes(x=log10(GMAC.Nominal.p), y=log10(Perm.p.GMAC.floored), color=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('Obs. Perm. P-value')),
       x = expression(log10('Obs. Nominal P-value')),
       subtitle = "GMAC Nominal vs Permutation")+
  scale_x_continuous(limits = c(-16,0))+
  scale_y_continuous(limits = c(-5,0))+
  geom_abline(intercept = 0, slope = 1, linetype="dotted")+
  geom_vline(xintercept = log10(h), color="red", linetype="dashed")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")

ggarrange(E, G, labels = c("E","G") , nrow=2, ncol=2)

```



## boxplots for PC inclusion differences

```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE, results='hide',fig.keep='all'}


sim$STM.med.p.floored=sim$STM.med.p
sim$STM.med.p.floored=replace(sim$STM.med.p.floored, which(sim$STM.med.p.floored<=2e-16), 2e-16)

sim$LTM.med.p.floored=sim$LTM.med.p
sim$LTM.med.p.floored=replace(sim$LTM.med.p.floored, which(sim$LTM.med.p.floored<=2e-16), 2e-16)

sim$TGM.med.floored=sim$TGM.med.p
sim$TGM.med.floored=replace(sim$TGM.med.floored, which(sim$TGM.med.floored<=2e-16), 2e-16)

TT=summary(as.factor(sim$STM.dim.diff))
rmv=as.numeric(names(TT[TT==1]))

sim.sub=sim[-c(match(rmv, sim$STM.dim.diff)),]

h=0.05
A=ggplot(data = sim.sub)+
  geom_boxplot(aes(x=as.factor(STM.dim.diff), y=log10(STM.med.p.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('STM Sim. Parametric P-value')),
       x = "Number of Extra Components In Analysis Model",
       subtitle = "Small True Model")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


B=ggplot(data = sim)+
  geom_boxplot(aes(x=as.factor(LTM.dim.diff), y=log10(LTM.med.p.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('LTM Sim. Parametric P-value')),
       x = "Number of Missing Components In Analysis Model",
       subtitle = "Large True Model")+
  geom_hline(yintercept = log10(h), color="red", linetype="dashed")


C=ggplot(data = sim)+
  geom_boxplot(aes(x=as.factor(Num.pcs.GMAC-Num.pcs.MRPC), y=log10(TGM.med.floored), fill=colvar))+
  theme(legend.position = "none")+
  labs(y = expression(log10('TGM Sim. Parametric P-value')),
       x = "Number of Missing Components In Analysis Model",
       subtitle = "True GMAC Model")+

  geom_hline(yintercept = log10(h), color="red", linetype="dashed")




ggarrange(A, B, C, labels = c("A","B","C") , nrow=3)


```






























## variation differences between sig and non-sig trios GMAC/MRPC



```{r, fig.cap="", echo=FALSE, message=FALSE, warning=FALSE,results='hide',fig.keep='all'}




set.seed(222)


h=0.05

sim.sub1=subset(sim, Med.pvalue.GMAC<0.05 & Med.pvalue.MRPC>0.05)
sim.sub1$`Model Significance`=rep("GMAC p < 0.05 vs MRPC p > 0.05", dim(sim.sub1)[1])
sim.sub2=subset(sim, Med.pvalue.GMAC>0.05 & Med.pvalue.MRPC>0.05)
sim.sub2$`Model Significance`=rep("GMAC p > 0.05 vs MRPC p > 0.05", dim(sim.sub2)[1])

sim.sub=rbind.data.frame(sim.sub1, sim.sub2)
sim.sub$`Model Significance`=as.factor(sim.sub$`Model Significance`)

A=ggplot(data = sim.sub)+
  geom_boxplot(aes(x=`Model Significance`, y=Per.var.GMAC-Per.var.MRPC, fill=`Model Significance`))+
  theme(axis.text.x=element_blank())+
  labs(y="Diff. In Cumulative Variation In Included PC's")

mean1=mean(sim.sub1$Per.var.GMAC-sim.sub1$Per.var.MRPC)
mean2=mean(sim.sub2$Per.var.GMAC-sim.sub2$Per.var.MRPC)

B=ggplot(data = sim.sub)+geom_density(aes(x=Per.var.GMAC-Per.var.MRPC, fill=`Model Significance`), alpha=0.3, show.legend = FALSE)+
  geom_vline(xintercept = mean1, linetype="dotted")+
  geom_vline(xintercept = mean2, linetype="dotted")+
  theme_minimal()+
  labs(x="Diff. In Cumulative Variation In Included PC's")


C=ggplot(data = sim)+
  geom_point(aes(x=GMAC.Nominal.p, y=Nominal.p.MRPC, color=colvar, label=Trio.Num))+
  theme(legend.position = "none")+
  labs(y = "Observed Nominal P-value (MRPC)",
       x = 'Observed Nominal P-value (GMAC)',
       subtitle = "MRPC vs. GMAC:  Color = Rare Variants")+
  geom_hline(yintercept = h, linetype="dashed", color="red")+
  #geom_vline(xintercept = h, linetype="dashed", color="red")+
  scale_y_continuous(limits = c(0, h))+
  geom_abline(slope = 1, intercept = 0, linetype="dashed")

  
D=ggplot(data = sim)+
  geom_point(aes(x=GMAC.Nominal.p, y=Nominal.p.MRPC, color=as.factor(ADDIS.inf.Class)))+
  theme(legend.position = "none")+
  labs(y = "Obs. Nominal P-value (MRPC)",
       x = 'Obs. Nominal P-value (GMAC)',
       subtitle = "MRPC vs. GMAC:  Color = Inf. Model Type")+
  geom_hline(yintercept = h, linetype="dashed", color="red")+
  geom_text(aes( -0.0004, h, label = h, vjust = -1), size = 5)+
  scale_y_continuous(limits = c(0, h))+
  geom_abline(slope = 1, intercept = 0, linetype="dashed")



ggarrange(A,B,C,D, labels = c("A","B","C","D"), nrow=2, ncol=2)


```