---
title: "Write-Up ADDIS Re-run and Verification"
subtitle: 'University of Idaho Department of Statistical Science'
author: "Jarred Kvamme"
date: '2/25/2021'
output:
  pdf_document: default
  html_document: default
  word_document: default
fontsize: 10pt
bibliography: biblies.bib
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=8)
```


\section*{Operations in January and February}

\subsection*{Re-analysis of GTEx Using {\bf MRPC-ADDIS}}

\textbf{DETAILED DESCRIPTION:}
Implementation of the ADDIS improvement to the MRPC algorithm: This involved adapting the existing scripts developed by M. Badsha to rerun the GTEx data using the ADDIS version of the FDR control. The distribution of model types (M0, M1, ... M4, Other) relative to each tissue was retained from the analyses for both \textbf{ MRPC-LOND} and \textbf{ MRPC-ADDIS} for comparison. Additionally, a host of programs were developed to identify the specific classification of each trio analyzed in each tissue. These programs were further adapted to identify the number of trios for each tissue classified as cis or trans mediated (M1 type 1 or M1 type 2). Such information was again compared with results from \textbf{MRPC LOND} with the intention of understanding the differences in inferred networks between the two FDR control methods.

In General, \textbf{MRPC-ADDIS} loosens the rejection threshold such that more edges/directions are inferred. There was typically a reshuffling of Model types with some specific trends such as most M1's whose graph classification was changed was converted to M2 or M4. M0's were converted to either M1, M2, M3, or M4 (with a large number in each tissue transferring to M3). The path by which M1 was converted to M2 or M4 was generally a direction flip between the cis and trans leading to a dependence structure of the cis gene dependent on both the trans gene and variant. The other method was the inference of an direct edge between the variant and both genes and a general edge between the cis and trans gene.   


\textbf{MAIN POINTS AND FILE LOCATIONS:}

$\bullet$ Re-run of GTEx data with \textbf{MRPC} using the \textbf{ADDIS} FDR control    
\indent $\rightarrow$ slightly modified version of M. Badsha's scripts.    
\indent $\rightarrow$ All associated stored data files in: 
$$/mnt/ceph/jarredk/AddisReRunFiles$$     
$\rightarrow$ The script used to Re-run \textbf{MRPC} using \textbf{ADDIS} is located at: 
$$/mnt/ceph/jarredk/ADDIS\_verify/ADDIS\_ReRun.R$$    



$\bullet$ Summarizing of differences between \textbf{LOND} and \textbf{ADDIS} trio classifications:   
$\rightarrow$ quantified the distribution of trio classifications for each tissue as counts of (M0, M1, ... M4, Other) 
$\rightarrow$ table can be found in:
$$/mnt/ceph/jarredk/ADDIS\_verify/Trios\_analysis\_GTEx\_v8\_allPCs\_V3\_ordered\_updated.xlsx$$   



$\bullet$ followed up on FDR methods by investigating specific trios and their classification differences between FDR methods:      

\textbf{GENERAL STEPS:}   
\begin{itemize}
  \item[1] read in SNP, PC, and final output files from $ADDIS\_rerun()$ 
  \item[2] catalog each trios \textbf{LOND} and \textbf{ADDIS} model classification
  \item[3] count/summarize trios between methods or within tissues (generally both)
  \item[4] rerun \text{MRPC} on trio using both \textbf{LOND} and \textbf{ADDIS} and compare graph and correlation matrices
\end{itemize}

$\rightarrow$ R-functions developed to assist in tracking trios are in: 
$$/mnt/ceph/jarredk/ADDIS\_verify/ADDIS\_Post\_Analysis\_processing.R$$



$\bullet$ followed up and compared mediation model trios of Type-1 (cis mediated) and Type-2 (trans-mediated) for both \textbf{ADDIS} and \textbf{LOND}   
$\rightarrow$ tables summarizing the counts and relative proportions of cis/trans mediated $M1$ models can be found at:
$$/mnt/ceph/jarredk/ADDIS\_verify/$$


```{r, fig.cap="Density plots showing the distribution of each graph type under both ADDIS and LOND", echo=FALSE, warning=FALSE, message=FALSE}
rraddis<- read.csv("C:/Users/Bruin/Desktop/Research Assistantship/ADDIS_ReRun/ReRunADDIS_Summary.csv")
rrlond<- read.csv("C:/Users/Bruin/Desktop/Research Assistantship/ADDIS_ReRun/ReRunLOND_Summary.csv")
nam=c("M0","M1","M2","M3","M4","Other")
modcts=rraddis[,c(6:11)]
colnames(modcts)=nam
modcts2=rrlond[,c(6:11)]
colnames(modcts2)=nam
modpcts=rraddis[,c(12:17)]
colnames(modpcts)=nam
modpcts2=rrlond[,c(12:17)]
colnames(modpcts2)=nam

L=rep("LOND", dim(modcts2)[1])
A=rep("ADDIS", dim(modcts)[1])

AplusL=c(A,L)

#pairs(modcts)
#pairs(modcts2)

plot.list=list()

# par(mfrow=c(2,3))
# for(i in 1:6){
#   hist(modcts[,i],main = paste("histogram of:", nam[i], sep = " "))
# }
# par(mfrow=c(1,1))




library(ggplot2)
for(i in 1:6){

  df=cbind.data.frame( FDR=as.factor(AplusL), M=c(modcts[,i], modcts2[,i]))
  
  plot.list[[i]]=ggplot(df, aes(x=M, color=FDR, fill=FDR)) +geom_density(alpha=0.3)+
    ggtitle(label = paste("Density plot for:", nam[i], sep = " "))
    
}


library(gridExtra)
grid.arrange(grobs=plot.list,nrows=3, ncol=2)
```






```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(knitr)

#change table
average.change.ct=colMeans(modcts-modcts2)
average.change.ct.sd=apply(modcts-modcts2, 2, sd)
average.change.pct=colMeans(modpcts-modpcts2)
average.change.pct.sd=apply(modpcts-modpcts2,2,sd)

change.table=rbind.data.frame(average.change.ct, average.change.ct.sd, 
                              average.change.pct, average.change.pct.sd)
colnames(change.table)=nam
row.names(change.table)=c("mean.change.ct", "SD.change.ct", "mean.change%", "SD.change%")

kable(change.table, caption = "table of the average difference in count and percentage between ADDIS and LOND and their standard errors for the differences", digits = 5)


#Bootstrap CI's

Addis.cIs=as.data.frame(matrix(0, nrow = 6, ncol = 3))
colnames(Addis.cIs)=c("lower.limit", "mean","upper.limit")
row.names(Addis.cIs)=nam
B=1000
alpha=0.05


#calculate confidence interval for each model type avg count(ADDIS)
for(i in 1:6){
  bsdata=rep(0,B)
  
  for(j in 1:B){
    
    bsdata[j]=mean(sample(modcts[,i], dim(modcts)[1], replace = TRUE))
    
  }
  
  #hist(bsdata)
  SE=sd(bsdata)
  M=mean(bsdata)
  lower=M-qnorm(1-alpha/2)*(SE/sqrt(length(bsdata)))
  upper=M+qnorm(1-alpha/2)*(SE/sqrt(length(bsdata)))
  
  Addis.cIs[i,]=round(c(lower,M,upper),3)
}



Lond.cIs=as.data.frame(matrix(0, nrow = 6, ncol = 3))
colnames(Lond.cIs)=c("lower.limit", "mean","upper.limit")
row.names(Lond.cIs)=nam
#calculate confidence interval for each model type avg count(LOND)
for(i in 1:6){
  bsdata=rep(0,B)
  
  for(j in 1:B){
    
    bsdata[j]=mean(sample(modcts2[,i], dim(modcts2)[1], replace = TRUE))
    
  }
  
  #print(bsdata)
  #hist(bsdata, breaks = 15)
  SE=sd(bsdata)
  M=mean(bsdata)
  lower=M-qnorm(1-alpha/2)*(SE/sqrt(length(bsdata)))
  upper=M+qnorm(1-alpha/2)*(SE/sqrt(length(bsdata)))
  
  Lond.cIs[i,]=round(c(lower,M,upper),3)
}


kable(Lond.cIs, caption = "Table of confidence intervals for the average count of each classified graph for all tissues under the LOND FDR control ")
kable(Addis.cIs, caption = "Table of confidence intervals for the average count of each classified graph for all tissues under the ADDIS FDR control ")

```

