---
title: 'Contribution Write-up 1/13/2021-2/26/2021'
subtitle: 'University of Idaho Department of Statistical Science'
author: "Jarred Kvamme"
date: '2/25/2021'
output:
  pdf_document: default
  html_document: default
  word_document: default
fontsize: 10pt
bibliography: biblies.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=8)
```

\section*{MAIN POINTS:}

\subsection*{(1) Identification of Trans Mediated Trios and Respective Meta-Data}    

$\bullet$ Mapping Quality thresholded chromatin interaction data for four tissues: Lung, Skin, Lymphoblastoid Cells, Fibroblast Cells  (See table 1).    

\begin{table}
\centering
\caption{File and dataset ID numbers for mapping quality thresholded chromatin interactions obtained from the ENCODE consortium }
\begin{tabular}{c|c|c}
\hline
\textbf{Tissue}      & \textbf{File ID}       & \textbf{Dataset ID}     \\ \hline
Lung                 & $ENCFF366ERB$          & $ENCSR645ZPH$           \\ \hline
Skin                 & $ENCFF569RJM$          & $ENCSR912RAV$           \\ \hline
Lymphoblastoid Cells & $ENCFF355OWW$          & $ENCSR423SYP$           \\ \hline
Fibroblast Cells     & $ENCFF768UBD$          & $ENCSR220MNC$           \\
\end{tabular}
\end{table}




$\bullet$ All trans-mediated (M1 type-2) classified trios from both \textbf{ADDIS} and \textbf{LOND} methods were extracted using \textit{ADDIS.M1.check()} function

$\bullet$ For each trio the SNP, Cis, and trans gene chromosome number and base pair positions on the chromosome were obtained for alignment with HiC data

\subsection*{(2) Extraction of HiC Interaction Data}
$\bullet$ We used the \textit{StrawR} package in $R$ to extract reads between specific regions on the SNP and trans-gene chromosomes. 

$\bullet$ The extracted reads are returned as counts per bin for all bins nested within the desired region and at a desired resolution. 

$\bullet$ In \textit{StrawR} resolution refers to the bin-size applied to the counts between chromosomal regions. Therefore the number of interactions for a region is the sum of the counts for all bins nested in that region

$\bullet$ The resolution was arbitrarily set to 10,000 base pairs which was the smallest resolution shared by all tissues.

$\bullet$ To ensure coverage of the regulatory elements of both the SNP and trans-gene, we extracted and totaled all interactions between 200,000 base pair regions on each chromosome and centered at the SNP's/trans-gene's starting position on that chromosome

$\bullet$ if no interactions existed for a SNP/trans-gene pair, the count/sum was replaced by "NA" to indicate the information was not available

$\bullet$ The count of interactions between the SNP and trans-gene was carried out and stored for each M1 type-2 trio classified by both \textbf{MRPC} methods

\subsection*{(3) Calculation of Empirical P-values By MC Sampling}

$\bullet$ Empirical p-values were used to determine if an observed number of interactions for a SNP/trans-gene pair was significant

$\bullet$ Empirical p-values were calculated by randomly sampling the interactions between the SNP chromosome and the trans-gene chromosome to construct a empirical null distribution of interactions (see section (4) for more detail)

$\bullet$ Two empirical p-values were calculated:
\begin{itemize}

	\item[(1)] the ratio of the number of non-NA samples larger than the observed number of interactions to of the total number of non-NA samples
	\item[(2)] the ratio of the number of non-NA samples larger than the observed number of interactions to the total number of samples

\end{itemize}


\subsection*{(4) MC Sampling Procedure}

$\bullet$ A single sample point was defined as the total number of interactions between two randomly positioned 200,000 bp regions; one on the SNP chromosome and one on the trans chromosome. 

$\bullet$ At each iteration in the sampling procedure, we first selected a single position on each chromosomes by a random draw from a uniform distribution with parameters tailored to the respective chromosome

$\bullet$ The randomly selected position was then treated as the center of a 200,000 bp region at that position. Again this was done for both the SNP and trans-gene chromosomes 

$\bullet$ each position was sampled without the possibility of replacement, but regions had the potential to overlap on smaller chromosomes

$\bullet$ the min/max parameters of the uniform distribution used to select a position on a given chromosome were set to be the minimum base pair number represented by the data plus 100,000 bp and the maximum base pair number represented by the data minus 100,000 bp

$\bullet$ the parameters were constrained to the above values so that any randomly selected 200,000 bp region would lie completely within the available data for the respective chromosome.

$\bullet$ this sampling process was repeated 10,000 times for each SNP/trans-gene pair to produce a unique empirical null distribution representative of the number of interactions between the SNP chromosome and trans-gene chromosome

\subsection*{(5) Multiple Comparisons Adjustments}

$\bullet$ Significant interaction enrichment was defined as an empirical p-value less than the threshold $\alpha$ taken at the usual level of $\alpha=0.05$

$\bullet$ To account for multiple testing we employed 3 correction techniques to the empirical p-values:

\begin{itemize}

	\item[(1)] the Holm-Bonferroni (HB) Family Wise Error Rate (FWER) adjustment 
	\item[(2)] Benjamin Yekutieli (BY) control of the false discovery rate
	\item[(3)] Benjamin Hochberg (BH) control of the false discovery rate

\end{itemize}

$\bullet$ the HB corrected p-values were compared at the original rejection threshold $\alpha = 0.05$

$\bullet$ The resulting q-values for the BH/BY methods were compared the desired false discovery rate $\beta=0.05$

$\bullet$ After applying the corrections, none of the empirical p-values remained significant
