

##########################################
#A fast example using fake genotype data
##########################################

#####################
#set the loading and saving locations and bp proximity
#####################
# specify the size of the neighborhood
bp.set.range=500000
# specify the path to the input files and path to the output files
filepath = "/mnt/ceph/jarredk/Methyl/Correlation_Code/"
input.location = paste (filepath, "input/", sep='')
output.location = paste (filepath, "output/", sep='')
# This example generates two output files: filename1 and filename2
# The filenames are automatically generated by function calc.corsV2 below 

# load functions in calcCOR_v2.R for correlation calculation
source(file = paste(input.location, "calcCOR_v2.R")) 


####################
#load in Data
# Note that only the methylation data file (filename) takes ~20 sec to load
# Other files are loaded quickly
####################
# genotype data: a data matrix of ??? individuals in the rows and ??? SNPs in the columns
genotype_data=read.table(file=paste(load.location, "EX_genotype_data.txt"), sep="\t", header=T)
# genotype metadata: ...
G_metadata=read.table(file = paste(load.location, "EX_Genotype_Metadata.txt", sep = ""), sep = "\t", header = T)
# ...
M_metadata=loadRData(fileName = paste(load.location, "meta_M.final.Rdata", sep = ""))
# ...
E_metadata=loadRData(fileName = paste(load.location, "meta_E.final.Rdata", sep = ""))
Edata=loadRData(fileName = paste(load.location, "Expression_data_BM_aligned.final.Rdata", sep = ""))
Mdata=loadRData(fileName = paste(load.location, "Mdata.final.Rdata", sep = ""))

######################
#Run for Chr 1
# Chr 1 is specified in the 'chrs' argument
######################
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("1"),
            fn=output.location,
            bp.range = bp.set.range)


