
loadRData <- function(fileName=NULL){
  #loads an RData file, and returns it
  load(fileName)
  get(ls()[ls() != "fileName"])
}

hm450_meta=read.csv(file="/mnt/ceph/jarredk/Methyl/GPL13534_HumanMethylation450_15017482_v.1.1.csv")
#gb37=read.table(file="/mnt/ceph/jarredk/Methyl/mart_export_GB37.txt", header = T, sep = ",")
#colnames(gb37) = c("Gene.name", "chr", "start.bp", "end.bp", "IlmnID")
rM=read.table("/mnt/ceph/jarredk/Methyl/CC_all/Correlation_CalculationV5_Scott/Example_Output/M_chr1_correlations.txt",sep="\t", header=T)
rE=read.table("/mnt/ceph/jarredk/Methyl/CC_all/Correlation_CalculationV5_Scott/Example_Output/E_chr1_correlations.txt",sep="\t", header=T)
GNs=loadRData(fileName = "/mnt/ceph/jarredk/Methyl/ExpressData/Express.BM.aligned.Ilmn.aligned.Genenames.Rdata")
Emeta=loadRData(file = "/mnt/ceph/jarredk/Methyl/ExpressData/Express.BM.aligned.Ilmn.aligned.MetaData.Rdata")

probes=unique(rM$Methyl_Probe_ID)

print(unique(hm450_meta[match(probes, hm450_meta$IlmnID),]$CHR))


E.ids=unique(rE$Express_Gene_ID)

#gb37.red=gb37[which(gb37$IlmnID %in% E.ids),]

#idx=which(gb37.red$Gene.name %in% GNs)

#print(unique(gb37.red[idx,]$chr))




Emeta.red=Emeta[which(Emeta$ID %in% E.ids),]

print(unique(Emeta.red$chr))














































































































































































































