############################################
# R script for calculating correlations
# for all the genotypes
# 
# Note: Please change the file path and provide the filenames for 
# the genotype data and the genotype meta data where necessary.
#
# Note: all input files must be in the same folder specified in /filepath/Input
# Also, output tables will go to the provided empty folder "Output" in the 
# /filepath/ 
#
# The script currently runs each chromosome sequentially, from chr1 to chr22, plus chrX and chrY.
# It will generate 2 output files for each chromosome in empty folder /Output/ 
# If genotype data are not available for chrX and chrY, please remove those two lines (starting at line 306)
# 
# Alternatively, this script may be duplicated for each chromosome 
# such that the calculation may be performed in parallel.
#
#############################################


#####################
#set the loading and saving locations and bp proximity
#####################
# specify the size of the neighborhood
bp.set.range=500000
# specify the path to the input files and path to the output files
filepath = "/mnt/ceph/jarredk/Methyl/Correlation_Code/"
input.location = paste (filepath, "Input/", sep='')
# the user must create the folder "Output" within filepath/
output.location = paste (filepath, "Output/", sep='')
# The filenames are automatically generated by function calcCOR_v2.R below 

# load functions in calcCOR_v2.R for correlation calculation
source(file = paste(filepath, "calcCOR_v2.R", sep = ""))

####################
#load in Data
# Note that only the methylation data file (Mdata.Rdata) takes ~30-60 sec to load
# Other files are loaded quickly
####################
# the user MUST fill in the filename1 and filename2 for the genotype data and genotype metadata files
# genotype data: a data matrix of 606 individuals and n SNPs in the columns
genotype_data=read.table(file = paste(input.location, "filename1", sep = ""), sep="\t", header=T)
# genotype metadata: a meta data matrix of 1000 SNPs in the rows and 2 columns [chr  coordinate]
G_metadata=read.table(file = paste(input.location, "filename2", sep = ""), sep = "\t", header = T)
# Methylation metadata: a meta data matrix of 363,516 probes and 3 columns [Ilmn_ID    chr    coordinate]
M_metadata=loadRData(fileName = paste(input.location, "meta_M.Rdata", sep = ""))
# Expression metadata: a meta data matrix of 20,578 probes in rows and 5 columns [gene_ID     chr    gene.start   gene.end   ILMN_ID]
E_metadata=loadRData(fileName = paste(input.location, "meta_E.Rdata", sep = ""))
# Expression Data: a matrix of 606 individuals in rows and 20,578 genes/probes in columns
Edata=loadRData(fileName = paste(input.location, "Expression_data_BM_aligned.Rdata", sep = ""))
# Methylation Data: a matrix of 606 individuals in rows and 363,516 probes in columns
Mdata=loadRData(fileName = paste(input.location, "Mdata.Rdata", sep = ""))


######################
#Run for each Chr
# Chr is specified in the 'chrs' argument
######################

calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("1"),
            fn=output.location,
            bp.range = bp.set.range)



#Run for Chr 2
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("2"),
            fn=output.location,
            bp.range = bp.set.range)


#Run for Chr 3
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("3"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 4
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("4"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 5
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("5"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 6
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("6"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 7
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("7"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 8
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("8"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 9
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("9"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 10
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("10"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 11
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("11"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 12
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("12"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 13
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("13"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 14
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("14"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 15
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("15"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 16
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("16"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 17
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("17"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 18
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("18"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 19
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("19"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 20
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("20"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 21
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("21"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr 22
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("22"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr X
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("X"),
            fn=output.location,
            bp.range = bp.set.range)

#Run for Chr Y
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("Y"),
            fn=output.location,
            bp.range = bp.set.range)


