

######################################################
#A fast example using fake genotype data (~100 SNPs)
######################################################

#####################
#set the loading and saving locations and bp proximity
#####################
# specify the size of the neighborhood
bp.set.range=500000
# specify the path to the input files and path to the output files
filepath = "/mnt/ceph/jarredk/Methyl/Correlation_Calculation/"
input.location = paste (filepath, "Input/", sep='')
output.location = paste (filepath, "Example_Output/", sep='')
# This example generates two output files: filename1 and filename2
# The filenames are automatically generated by function calc.corsV2 below 

# load functions in calcCOR_v2.R for correlation calculation
source(file = paste(filepath, "calcCOR_v2.R", sep = ""))

####################
# load in Data
# Note that only the methylation data file (Mdata.Rdata) takes ~30 sec to load
# Other files are loaded quickly
####################
# genotype data: a data matrix of 606 individuals in the rows and 1000 SNPs in the columns
genotype_data=read.table(file = paste(input.location, "Example_Genotype_data.txt", sep = ""), sep="\t", header=T)
# genotype metadata: a meta data matrix of 1000 SNPs in the rows and 3 columns [SNP_ID    chr  coordinate]
G_metadata=read.table(file = paste(input.location, "Example_Genotype_Metadata.txt", sep = ""), sep = "\t", header = T)
# Methylation metadata: a meta data matrix of 363,516 probes and 3 columns [Ilmn_ID    chr    coordinate]
M_metadata=loadRData(fileName = paste(input.location, "meta_M.Rdata", sep = ""))
# Expression metadata: a meta data matrix of xxx probes and 3 columns [gene_ID     chr       coordinate]
E_metadata=loadRData(fileName = paste(input.location, "meta_E.Rdata", sep = ""))
# Expression Data: a matrix of 606 individuals in rows and xxx genes/probes in columns
Edata=loadRData(fileName = paste(input.location, "Expression_data_BM_aligned.Rdata", sep = ""))
# Methylation Data: a matrix of 606 individuals in rows and 363,516 probes in columns
Mdata=loadRData(fileName = paste(input.location, "Mdata.Rdata", sep = ""))

######################
#Run for Chr 1
# Chr 1 is specified in the 'chrs' argument
######################
calc.corsV2(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c("1"),
            fn=output.location,
            bp.range = bp.set.range)


