############################################
# R script for calculating correlations
# for all the genotypes
# 
# Note: Please provide the filenames for
# the genotype data and the genotype metadata in lines 44 and 46.
#
# All input files are in ./Input/.
# Output files will go to the provided empty folder ./Output/.
#
# The script currently runs on each chromosome sequentially,
# from chr1 to chr22, plus chrX and chrY.
# It will generate 2 output files for each chromosome in empty folder ./Output/.
# If genotype data are not available for chrX and chrY,
# please remove those two lines (starting at line 306).
# 
# Alternatively, this script may be duplicated for each chromosome 
# such that the calculation may be performed in parallel.
#
#############################################


#####################
# Set the loading and saving locations
#####################
# specify the path to the input files and path to the output files
#filepath = "/mnt/ceph/jarredk/Methyl/Correlation_Calculation/"
filepath = "./"
input.location = paste (filepath, "Input/", sep='')
# the user must create the folder "Output" within filepath/
output.location = paste (filepath, "Output/", sep='')
# The filenames are automatically generated by function calcCOR_v2.R below 

# load functions in calcCOR_v2.R for correlation calculation
source(file = paste(filepath, "calcCOR_v5a.R", sep = ""))

####################
# Load in data
#
# Note that only the methylation data file (Mdata.Rdata) takes ~30-60 sec to load
# Other files are loaded quickly
####################
# Please fill in the filename1 and filename2 for the genotype data and genotype metadata files
# genotype data: a data matrix with SNPs in the rows and individuals in the columns
print('loading genotype data')
genotype_data=read.table(file = gzfile(paste(input.location, "Genotypes_additive_genotypedata.txt.gz", sep = "")), sep="\t", header=T)
# genotype metadata: a data matrix SNPs in the rows
# and 3 columns [SNP_ID, chr, coordinate, CountedAllele, OtherAllele]
print('loading genotype metadata')
G_metadata=read.table(file = gzfile(paste(input.location, "Genotypes_additive_genotypemetadata.txt.gz", sep = "")), sep = "\t", header = T)

# [QIMR variation] convert PLINK-style chromosome numbers to their normal 'text' names (chr X, X/Y, Y, MT) - otherwise the correlations for the X chromosome don't work. This still doesn't rescue the Y chromosome.
G_metadata$chr[G_metadata$chr=='23'] = 'X'
G_metadata$chr[G_metadata$chr=='24'] = 'Y'
G_metadata$chr[G_metadata$chr=='25'] = 'XY'
G_metadata$chr[G_metadata$chr=='26'] = 'MT'

# Methylation metadata: a meta data matrix of 363,516 probes
# and 3 columns [Probe_ID, chr, coordinate]
print('loading methylation metadata')
M_metadata=loadRData(fileName = paste(input.location, "All.Mprobes.Metadata.Rdata", sep = ""))
# Methylation Data: a matrix of 606 individuals in rows and 363,516 probes in columns
print('loading methylation data')
Mdata=loadRData(fileName = paste(input.location, "MethylData.RegressResids.asMatrix.Rdata", sep = ""))

# Expression metadata: a meta data matrix of 20,578 probes in rows
# and 5 columns [gene_ID, chr, gene.start, gene.end, ILMN_ID]
print('loading expression metadata')
E_metadata=loadRData(fileName = paste(input.location, "Express.BM.aligned.Ilmn.aligned.MetaData.Rdata", sep = ""))
# Expression Data: a matrix of 606 individuals in rows and 20,578 genes/probes in columns
print('loading expression data')
Edata=loadRData(fileName = paste(input.location, "Express.BM.aligned.Ilmn.aligned.Rdata", sep = ""))


######################
# Run for each Chr
# Chr is specified in the 'chrs' argument
######################
# specify the size of the neighborhood
bp.set.range=500000

for (chr in c(1:22,'X','Y'))
#for (chr in c('X','Y'))
{ print(paste(sep='','calculating correlations, chromosome ',chr))
  calc.cors(mmat = Mdata,
            emat = Edata,
            gmat = genotype_data, 
            GMInfo = M_metadata, 
            GEInfo = E_metadata, 
            genoInfo = G_metadata, 
            chrs=c(chr),
            fn=output.location,
            bp.range = bp.set.range)
}


